smolyay/basis.py

```diff
@@ -1,34 +1,81 @@
 import abc
-import math
 
 import numpy
 import scipy.special
 
+
 class BasisFunction(abc.ABC):
     """Basis function for interpolating data.
 
-     A one-dimensional basis function is defined on the domain
-     :math:`[-1,1]`. The function defines the :attr:`points` at
-     which it should be sampled within this interval for interpolation.
-     The function also has an associated :meth:`__call__` method
-     for evaluating it at a point within its domain. Moreover,
-     the first derivative of the function can be evaluated via
-     :meth:`derivative`.
+    A one-dimensional basis function is defined on some given natural
+    domain. The function defines the :attr:`points` at which it should
+    be sampled within this interval for interpolation. The function also
+    has an associated :meth:`__call__` method for evaluating it at a
+    point within its domain. Moreover, the first derivative of the function
+    can be evaluated via :meth:`derivative`.
     """
 
-    def __init__(self):
-        self._points = []
-
     @property
     @abc.abstractmethod
-    def points(self):
-        """list: Sampling points for interpolation."""
+    def domain(self):
+        """list: Domain the sample points come from."""
         pass
 
-    @abc.abstractmethod
     def __call__(self, x):
         """Evaluate the basis function.
 
+        Parameters
+        ----------
+        x : float
+            One-dimensional point.
+
+        Returns
+        -------
+        float
+            Value of basis function.
+        """
+        if not numpy.all(self.check_in_domain(x)):
+            raise ValueError("Input is outside the domain " + str(self.domain))
+        return self._function(x)
+
+    def derivative(self, x):
+        """Evaluate the first derivative of the basis function.
+
+        Parameters
+        ----------
+        x : float
+            one-dimensional point.
+
+        Returns
+        -------
+        float
+            Value of the derivative of the basis function.
+        """
+        if not numpy.all(self.check_in_domain(x)):
+            raise ValueError("Input is outside the domain")
+        return self._derivative(x)
+
+    def check_in_domain(self, x):
```

Comment:
I know I suggested the original name for this, but seeing how you're using it, I think `in_domain` might be better:
```suggestion
    def in_domain(self, x):
```